File: DHG.py
import requests 
import json

TOKEN = "your-api-key"
HEADERS = {
    "Authorization": f"Token {TOKEN}",
    "Accept": "application/vnd.github.v3+json",
}

dihypergraph = {
    "nodes": [],
    "hyperedges": []
}

class DHGConstructor:
    def __init__(self, BASE_URL, HEADERS, dhg_name, init_empty = True, dhg = None):
        self.dhg_name = dhg_name
        self.issues = requests.get(BASE_URL + "issues", headers = HEADERS).json()
        self.commits = requests.get(BASE_URL + "commits", headers = HEADERS).json()
        self.prs = requests.get(BASE_URL + "pulls", headers = HEADERS).json()

        self.BASE_URL = BASE_URL
        self.HEADERS = HEADERS

        if init_empty == True:
            self.dihypergraph = {
                "nodes": [],
                "hyperedges": []
            }
        else:
            self.dihypergraph = dhg

    def run_loop(self):
        for issue in self.issues:
            self.dihypergraph["nodes"].append({
                "id": f"issue{issue['id']}",
                "type": "issue",
                "data": issue
            })
            self.dihypergraph["hyperedges"].append({
                "source": [f"user_{issue['user']['id']}"],
                "target": [f"issue_{issue['id']}"],
                "type": "created_by",
                "data": {}
            })
        
        for commit in self.commits:
            self.dihypergraph["nodes"].append({
                "id": f"commit_{commit['sha']}",
                "type": "commit",
                "data": commit
            })
        if commit['committer']:
            dihypergraph["hyperedges"].append({
                "source": [f"user_{commit['committer']['id']}"],
                "target": [f"commit_{commit['sha']}"],
                "type": "committed_by",
                "data": {}
            })
        else:
            pass

        for pr in self.prs:
            self.dihypergraph["nodes"].append({
                "id": f"pushRequest_{pr['head']['sha']}",
                "type" : "pushRequest",
                "data": pr
            })
        
        with open(f"dihypergraph_{self.dhg_name}.json", "w") as outjson:
            json.dump(self.dihypergraph, outjson, indent = 4)
        print(f"Dihypergraph constructed to dihypergraph_{self.dhg_name}")
                                
if __name__ == "__main__":
    BASE_URL = "https://api.github.com/repos/plurigrid/ontology/"
    TOKEN = "your-api-token"
    HEADERS = {
        "Authorization": f"token {TOKEN}",
        "Accept": "application/vnd.github.v3+json",
    }

    my_dhg = DHGConstructor(BASE_URL, HEADERS, "my_sample")
    my_dhg.run_loop()

File: DHG-clj/DHG.clj
#!/usr/bin/env bb

(defn read-api-key [file-path]
  (slurp file-path))

(defn fetch-data [url headers]
  (:body (clojure.java.io/as-url url)
         {:headers headers :as :json}))

(defn create-nodes [issues commits prs]
  (concat
   (map #(hash-map :id (str "issue" (:id %)) :type "issue" :data %) issues)
   (map #(hash-map :id (str "commit_" (:sha %)) :type "commit" :data %) commits)
   (map #(hash-map :id (str "pushRequest_" (get-in % [:head :sha])) :type "pushRequest" :data %) prs)))

(defn create-hyperedges [issues commits]
  (concat
   (map #(hash-map :source [(str "user_" (get-in % [:user :id]))]
                   :target [(str "issue_" (:id %))]
                   :type "created_by"
                   :data {}) issues)
   (map #(hash-map :source [(str "user_" (get-in % [:committer :id]))]
                   :target [(str "commit_" (:sha %))]
                   :type "committed_by"
                   :data {}) commits)))

(defn construct-dhg [base-url headers dhg-name]
  (let [issues   (fetch-data (str base-url "issues") headers)
        commits  (fetch-data (str base-url "commits") headers)
        prs      (fetch-data (str base-url "pulls") headers)
        nodes    (create-nodes issues commits prs)
        hyperedges (create-hyperedges issues commits)]
    {:nodes nodes :hyperedges hyperedges}))

(defn save-dhg [dhg dhg-name]
  (spit (str "dihypergraph_" dhg-name ".json") (with-out-str (prn dhg))))

(defn main []
  (let [base-url "https://api.github.com/repos/kennethZhangML/StochFlow/"
        token (read-api-key "/Users/kennethzhang/Desktop/gh-apikey.txt")
        headers {"Authorization" (str "token " token)
                 "Accept" "application/vnd.github.v3+json"}
        dhg-name "stochflow"
        dhg (construct-dhg base-url headers dhg-name)]
    (save-dhg dhg dhg-name)
    (println (str "Dihypergraph constructed to dihypergraph_" dhg-name))))

(main)
